infrastructures/docker/src/main/java/org/eclipse/che/workspace/infrastructure/docker/Bootstrapper.java

```diff
@@ -0,0 +1,185 @@
+    private final AtomicInteger                              endpointId;
+    private final EventSubscriber<BootstrapperStatusEvent>   bootstrapperStatusListener;
+    private       CompletableFuture<BootstrapperStatusEvent> finishEventFuture;
+
+    @Inject
+    public Bootstrapper(@Assisted String machineName,
+                        @Assisted RuntimeIdentity runtimeIdentity,
+                        @Assisted DockerMachine dockerMachine,
+                        @Named("che.workspace.che_server_ws_endpoint_base") String wsBaseEndpoint,
+                        @Named("che.docker.bootstrapping_timeout_min") int bootstrappingTimeoutMinutes,
+                        EventService eventService) {
+        this.machineName = machineName;
+        this.runtimeIdentity = runtimeIdentity;
+        this.dockerMachine = dockerMachine;
+        this.bootstrappingTimeoutMinutes = bootstrappingTimeoutMinutes;
+        this.wsIntallerEndpoint = wsBaseEndpoint + InstallerEndpoint.INSTALLER_WEBSOCKET_ENDPOINT_BASE;
+        this.eventService = eventService;
+        this.endpointId = new AtomicInteger();
+        this.bootstrapperStatusListener = event -> {
+            BootstrapperStatus status = event.getStatus();
+            //skip starting status event
+            if (status.equals(BootstrapperStatus.DONE) || status.equals(BootstrapperStatus.FAILED)) {
+                //check boostrapper belongs to current runtime and machine
+                RuntimeIdentityDto runtimeId = event.getRuntimeId();
+                if (event.getMachineName().equals(machineName)
+                    && runtimeIdentity.getEnvName().equals(runtimeId.getEnvName())
+                    && runtimeIdentity.getOwner().equals(runtimeId.getOwner())
+                    && runtimeIdentity.getWorkspaceId().equals(runtimeId.getWorkspaceId())) {
+
+                    finishEventFuture.complete(event);
+                }
+            }
+        };
+    }
+
+    /**
+     * Bootstraps installers and wait while they finished.
+     *
+     * @param agents
+     *         agents to install
+     * @throws InfrastructureException
+     *         when bootstrapping timeout reached
+     * @throws InfrastructureException
+     *         when bootstrapping failed
+     * @throws InfrastructureException
+     *         when any other error occurs while bootstrapping
+     */
+    public void bootstrap(List<AgentImpl> agents) throws InfrastructureException {
+        if (finishEventFuture != null) {
+            throw new IllegalStateException("Bootstrap method must be called only once.");
+        }
+        this.finishEventFuture = new CompletableFuture<>();
+
+        injectBootstrapper(agents);
+
+        this.eventService.subscribe(bootstrapperStatusListener, BootstrapperStatusEvent.class);
+        try {
+            String endpoint = wsIntallerEndpoint + endpointId.getAndIncrement();
+            dockerMachine.exec(BOOTSTRAPPER_DIR + BOOTSTRAPPER_FILE +
```

Comment:
Nice catch. I missed these parameters. I'll add configuration properties for them.